### ISSUE TITLE
New Feature - enable use of reduction privileges with the privilege calculations.

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/RegentParticleDSL#45]
At the moment we use only writes and reads privileges. This applies ordering/reduces parallelism between tasks on the same group of particles.

A more advanced option is to enable the reduction privileges in region, e.g.:
```
reduces+(r)
```

I believe that any field used in a `reduces` privilege cannot appear in a `reads` or `writes`, so we would need roughly the following:

1. A way to work out whether a write is a reduction operation or not. This would require more complex analysis of the AST, looking for expressions with the format or `a = a OPERATOR b`. We could initially limit to +/- operators as well. At a guess something like `if lhs = rhs.lhs` or something and then check the operator we could perhaps assume reduction is ok.
2. A way to actually determine if a field is actually read from or not (as opposed to the current method of all fields that appear are read from since we're assuming writes are a read with additional requirements anyway). This would mean checking all assignments/conditionals/etc. to check where the field appears. Furthermore, if the expression appears as a reduction we exclude it from the read (N.B. I'm not actually clear this works, needs to be tested).
3. Non-reduction operations are still writes as before.
4. Finally, check if anything we have that has a reduction privilege appears in a stronger (i.e. read or write privilege) clause still, and if so remove it from the reduction privilege list before returning.

This is some way off, but could improve performance in the future.

**Comment by LonelyCat124** [comment]
Note, use of atomic coherence mode may remove the requirement for this (http://regent-lang.org/reference/#coherence-modes)

**Comment by LonelyCat124** [comment]
So the reduction requirement is `+=` (or related). I'll look at how to read this imminently I think.

**Comment by LonelyCat124** [comment]
Ok, so the basis is that when we have a statement that is reducable, e.g.
```
space.field += space2.field - 3 * r2
```
or similar, then we get a `ast.specialized.stat.Reduce` node. The (important) members of this are:
`lhs[1]` - a FieldAccess for the left-hand-side field
`op` - the operator on which the reduction takes place.
`rhs` - the right hand side of the equation.

The current code is difficult to use correctly as right now all `ast.specialized.expr.FieldAccess` nodes are treated the same (i.e. read from), which means the lhs variable also appears in the reads field, and I don't believe a variable can appear in both a reads (or writes) privilege and a reduction privilege at the same time.

The ideal here is to not recurse from `ast.specialized.stat.Reduce` from the main recursion loop, it would be better to hit the `privilege_map` function, and then have that recurse on the `rhs` of the node. I'm not entirely sure how I should do that right now but I'll try a  few things out and see.

Regarding the generation of reduction privileges, I'm not 100% but I believe the call needed is ` regentlib.privilege(regentlib.reduces(OP), region, field`

**Comment by LonelyCat124** [comment]
Ok, so my current solution involves a few steps.

First I create two tables to determine which parts of the AST are worked on
```
local is_specialized_table = {
  [ast.specialized.stat.Reduce] = return_false,
  [ast.specialized.expr] = return_true,
  [ast.specialized.stat] = return_true,
  [ast.specialized.top] = return_true,
  [ast.specialized.Block] = return_true,
}

local is_spec_reduce_table = {
  [ast.specialized.expr.FieldAccess] = return_false,
  [ast.specialized.stat.Assignment] = return_false,
  [ast.specialized.expr] = return_true,
  [ast.specialized.stat] = return_true,
  [ast.specialized.top] = return_true,
  [ast.specialized.Block] = return_true,
}
```

The first does not do the recursion on the `Reduce` nodes, while the second does not do the recursion on the `FieldAccess` and `Assignment` nodes. These are the 3 node types we currently care about in the privilege map.

The traversal function is then changed to loop over the nodes twice, once for each of these limits:
```
  ast.traverse_node_postorder(
    function(node)
        fn(node, sym1, sym2, read_sym1, read_sym2, write_sym1, write_sym2)
    end,
    node, is_specialized_node)
  ast.traverse_node_postorder(
    function(node)
      fn(node, sym1, sym2, read_sym1, read_sym2, write_sym1, write_sym2)
    end,
    node, is_specialized_reduc_node)
```

The first loop will find all of the non-reduction FieldAccesses, and pull them out for reads/writes.
The latter, will find all of the reduction accesses, pull out the for the reduction, and then recurse on the RHS to find other FieldAccesses needed for the reads.

I guess this might be where its worth asking @sergisiso for if there's a better way to navigate the AST than two loops? I need to find all of the FieldAccesses (other than those on the left-hand side of a Reduction), as well as correctly identifying those on the left-hand side of an Assignment. I assume I could do something smarter but I'm a bit out of my depth for a "good" solution.

**Comment by LonelyCat124** [comment]
Next question is what the rules on variables that appear in reduction clauses is. My understanding would be "Fields that appear in reduction clauses cannot appear in read or write clauses.", but I'm not sure whether Regent has this requirement, so I'll check.

**Comment by LonelyCat124** [comment]
Ok, apparently Regent has no restriction on how the fields can appear in the clauses, so I don't need to worry overly about that.

I imagine if a field appears in both a read/write clause and a reduce clause there could be performance/serialisability concerns, so I think I'll try to detect that and print a warning, but not forbid it for now.

**Comment by LonelyCat124** [comment]
I have a small implementation of this working, so I'll expand that to the main code and update various implemented kernels to make use of reductions this week.

**Comment by sergisiso** [comment]
@LonelyCat124 In the PSyclone walk method we allow to match multiple node types in the same traversal which looks nice. But I agree with what you said later: if the order in which things are returned is not important then this is just a tool performance issue and not a correctness issue, so I wouldn't worry too much about it and can always be improved later.

**Comment by LonelyCat124** [comment]
@sergisiso Yeah its definitely only a performance issue - the difficulty I had doing it in one loop is that I look for both `FieldAccess` and `Reduction` nodes (to make my life easy), however as a child node the `Reduction` node contains a few `FieldAccess` nodes, and I don't want to treat those all the same (as they're a specific case). I think I'll leave it for now as it (seems to) works.

**Comment by LonelyCat124** [comment]
This is now working as far as I can tell. Not yet tested performance.

**Comment by LonelyCat124** [comment]
This now exists and is merged in #77
